File: Week3/hw3_var2.py
import numpy as np
from numpy import (pi, exp, sin, cos)
import matplotlib.pyplot as plt
from time import perf_counter
import os


# 1D Heat conductivity equation
# Homework 3, Problem 2

# Initial condition
# U(t=0, x) = 1 - x^3

# Boundary conditions
# Ux(t, x=0) = 0
# U(t, x=1) = 0

def Analytical_solution(x, t, m=100):
    λ = lambda n: (2*n - 1)*pi/2
    A_n = lambda n: -12*((-1)**n + 1/λ(n)) / λ(n)**3
    
    F = 0
    for n in range(1, m+1):
        F += A_n(n) * exp(-λ(n)**2*t) * cos(λ(n)*x)
    
    return F

def Simple_iterative_method(U:np.ndarray, N, dt, dx, a2=1, 
                                eps=1e-6, stop_iteration=3e4):
    U_old = U.copy()
    U_new = np.zeros_like(U)

    set_boundary_U(U=U_old)

    iteration = 0
    maximum = 1
    while maximum > eps and iteration < stop_iteration:
        U_new[1:N-1] = U_old[1:N-1] + a2 * dt / dx**2 \
            * (U_old[2:N] - 2*U_old[1:N-1] + U_old[0:N-2])
        
        set_boundary_U(U=U_new)
        
        maximum = np.max(np.abs(U_new - U_old))
        U_old = U_new.copy()
        iteration += 1

    return U_new, iteration

def set_boundary_U(U):
    U[0] = U[1]
    U[N-1] = 0

def Thomas_algorithm(U:np.ndarray, N, dt, dx, a2=1, 
                        eps=1e-6, stop_iteration=3e4):
    U_old = U.copy()
    U_new = np.zeros_like(U)

    alpha = np.zeros_like(U)
    beta = np.zeros_like(U)

    A = -a2 / dx**2
    B = 1 / dt + 2*a2 / dx**2
    C = -a2 / dx**2
    D = U_old / dt # numpy 1d array

    iteration = 0
    maximum = 1
    while maximum > eps and iteration < stop_iteration:
        D = U_old / dt
        alpha[1] = 1
        beta[1] = 0
        for i in range(1, N-1):
            alpha[i+1] = -A / (B + C*alpha[i])
            beta[i+1] = (D[i] - C*beta[i]) / (B + C*alpha[i])
        
        U_new[N-1] = 0
        for i in range(N-2, -1, -1):
            U_new[i] = alpha[i+1]*U_new[i+1] + beta[i+1]

        maximum = np.max(np.abs(U_new - U_old))
        U_old = U_new.copy()
        iteration += 1

    return U_new, iteration

dx = 0.1
dt = 0.001

start_x, end_x = (0, 1)
N = int((end_x - start_x) / dx) + 1
x = start_x + np.arange(start=0, stop=N) * dx

a2 = 1
eps = 1e-6
stop_iteration = 3e4

U_old = np.zeros_like(x)

# Initial condition
U_old[:] = 1 - x**3

start_time = perf_counter()
U_simple, iter_S = Simple_iterative_method(U_old, N, dt, dx, a2=1)
F_S = Analytical_solution(x, t=iter_S*dt)
end_time = perf_counter()

print("Simple Iterative Method")
print(f"Calculation time: {end_time - start_time:.6f} seconds")
print(f"Maximum error: {np.max(np.abs(F_S - U_simple)):e}")
print(f"Number of iterations: {iter_S}")

start_time = perf_counter()
U_Thomas, iter_T = Thomas_algorithm(U_old, N, dt, dx, a2=1)
F_T = Analytical_solution(x, t=iter_T*dt)
end_time = perf_counter()

print("Thomas algorithm")
print(f"Calculation time: {end_time - start_time:.6f} seconds")
print(f"Maximum error: {np.max(np.abs(F_T - U_simple)):e}")
print(f"Number of iterations: {iter_T}")

path = "Results"
# Create folder where we will store solution
if not os.path.exists(path):
    os.makedirs(path)

np.savetxt(f"{path}\\HW3_X_py.txt", x, fmt="%.6e", delimiter="\t")
np.savetxt(f"{path}\\HW3_U_Simple_py.txt", U_simple, fmt="%.6e", delimiter="\t")
np.savetxt(f"{path}\\HW3_F_Simple_py.txt", F_S, fmt="%.6e", delimiter="\t")
np.savetxt(f"{path}\\HW3_U_Thomas_py.txt", U_simple, fmt="%.6e", delimiter="\t")
np.savetxt(f"{path}\\HW3_F_Thomas_py.txt", F_T, fmt="%.6e", delimiter="\t")

plt.title("1D Heat conductivity equation")
plt.grid()
plt.plot(x, U_simple, label="Simple iterative method")
plt.plot(x, F_S, ls="--", label=f"Analytical solution at {dt*iter_S:.3f}")
plt.plot(x, U_Thomas, label="Thomas algorithm")
plt.plot(x, F_T, ls="--", label=f"Analytical solution at {dt*iter_T:.3f}")
plt.xlabel("x")
plt.ylabel("U(x)")
plt.legend()
plt.tight_layout()
plt.show()

File: Week3/hw3_var2.cpp
#include <iostream>
#include <cmath>
#include <chrono>
using namespace std::chrono;


// 1D Heat conductivity equation
// Homework 3, Problem 2

// Initial condition
// U(t=0, x) = 1 - x^3

// Boundary conditions
// Ux(t, x=0) = 0
// U(t, x=1) = 0

const double start_x = 0, end_x = 1;
const double dx = 0.1, dt = 0.001;
const double pi = 3.14159265359;
const double eps = 1e-6;
const double a2 = 1;

const int N = (end_x - start_x) / dx + 1;
const int stop_iteration = 3e5;

double x[N], F_S[N], F_T[N], U_old[N], U_S[N], U_T[N];

void FillArray(double arr[N], double value=0){
    for(int i = 0; i < N; ++i){
        arr[i] = value;
    }
}

double FindMaxDiff(double U_old[N], double U_new[N]){
    double maximum = 0;
    for(int i = 0; i < N; ++i){
        if(maximum < fabs(U_new[i] - U_old[i])){
            maximum = fabs(U_new[i] - U_old[i]);
        }
    }
    return maximum;
}

void Analytical_solution(double F[N], double t, int m=100){
    for(int n = 1; n <= m; ++n){
        for(int i = 0; i < N; ++i){
            F[i] += -96*(pow(-1, n) + 2/((2*n - 1)*pi))
                / pow((2*n - 1)*pi, 3)
                * exp(-pow((2*n - 1)*pi/2, 2)*t)
                * cos((2*n - 1)*pi/2*x[i]);
        }
    }    
}

void Simple_Iterative_Method(double U_old[N], double U_new[N], int N, double dt, 
                                double dx, double a2, double eps=1e-6, int stop_iteration=3e4){
    for(int i = 1; i < N-1; ++i){
        U_new[i] = U_old[i] + a2 * dt / (dx*dx)
            * (U_old[i+1] - 2*U_old[i] + U_old[i-1]);
    }
}

void set_boundary_U(double U[N]){
    U[0] = U[1];
    U[N-1] = 0;
}

void Thomas_algorithm(double U_old[N], double U_new[N], int N, double dt, 
                        double dx, double a2, double eps=1e-6, int stop_iteration=3e4){

    double alpha[N], beta[N], D[N];

    double A = -a2 / (dx*dx);
    double B = 1 / dt + 2*a2 / (dx*dx);
    double C = -a2 / (dx*dx);
    
    for(int i = 0; i < N; ++i){
        D[i] = U_old[i] / dt;
    }

    alpha[1] = 1;
    beta[1] = 0;
    for(int i = 1; i < N-1; ++i){
        alpha[i+1] = -A / (B + C*alpha[i]);
        beta[i+1] = (D[i] - C*beta[i]) / (B + C*alpha[i]);
    }

    U_new[N-1] = 0;
    for(int i = N-2; i >= 0; --i){
        U_new[i] = alpha[i+1]*U_new[i+1] + beta[i+1];
    }
}

int main(){
    for(int i = 0; i < N; ++i){
        x[i] = start_x + i*dx;
    }

    // Initialize arrays
    FillArray(U_old, 0);
    FillArray(U_S, 0);
    FillArray(U_T, 0);
    FillArray(F_S, 0);
    FillArray(F_T, 0);

    // Simple Iterative method
    // initial condition
    // U(x, t=0) = 1 - x^3
    for(int i = 0; i < N; ++i){
        U_old[i] = 1 - pow(x[i], 3);
    }

    auto start_S = high_resolution_clock::now();

    int iter_S = 0;
    double maximum = 0;
    do{
        Simple_Iterative_Method(U_old, U_S, N, dt, dx, a2);
        set_boundary_U(U_S);

        maximum = FindMaxDiff(U_old, U_S);

        for(int i = 0; i < N; ++i){
            U_old[i] = U_S[i];
        }

        iter_S++;
    }while(maximum > eps and iter_S < stop_iteration);

    Analytical_solution(F_S, dt*iter_S);

    maximum = FindMaxDiff(F_S, U_S);

    auto stop_S = high_resolution_clock::now();
    auto duration_S = duration_cast<microseconds>(stop_S - start_S);
    
    printf("calculating time: %.6f seconds\n", duration_S.count() / 1e6);
    printf("Maximum error (Simple): %e\n", maximum);
    printf("Number of iterations: %d\n", iter_S);

    // Thomas algorithm
    // initial condition
    // U(x, t=0) = 1 - x^3
    for(int i = 0; i < N; ++i){
        U_old[i] = 1 - pow(x[i], 3);
        U_T[i] = 0;
    }

    set_boundary_U(U_old);
    set_boundary_U(U_T);

    auto start = high_resolution_clock::now();

    int iter_T = 0;
    maximum = 0;
    do{
        Thomas_algorithm(U_old, U_T, N, dt, dx, a2);
        
        maximum = FindMaxDiff(U_old, U_T);

        for(int i = 0; i < N; ++i){
            U_old[i] = U_T[i];
        }

        iter_T++;
    }while(maximum > eps and iter_T < stop_iteration);

    Analytical_solution(F_T, dt*iter_T);

    maximum = FindMaxDiff(F_T, U_T);

    auto stop = high_resolution_clock::now();
    auto duration = duration_cast<microseconds>(stop - start);
    
    printf("calculating time: %.6f seconds\n", duration.count() / 1e6);
    printf("Maximum error (Thomas): %e\n", maximum);
    printf("Number of iterations: %d\n", iter_T);

    FILE *File;
    File = freopen("Results\\HW3_X_cpp.txt", "w", stdout);
    for(int i = 0; i < N; ++i){
        if(i == N-1){
            printf("%.9f", x[i]);
        }else{
            printf("%.9f\t", x[i]);
        }
    }
    fclose(File);

    File = freopen("Results\\HW3_U_Simple_cpp.txt", "w", stdout);
    for(int i = 0; i < N; ++i){
        if(i == N-1){
            printf("%.9f", U_S[i]);
        }else{
            printf("%.9f\t", U_S[i]);
        }
    }
    fclose(File);

    File = freopen("Results\\HW3_F_Simple_cpp.txt", "w", stdout);
    for(int i = 0; i < N; ++i){
        if(i == N-1){
            printf("%.9f", F_S[i]);
        }else{
            printf("%.9f\t", F_S[i]);
        }
    }
    fclose(File);

    File = freopen("Results\\HW3_U_Thomas_cpp.txt", "w", stdout);
    for(int i = 0; i < N; ++i){
        if(i == N-1){
            printf("%.9f", U_T[i]);
        }else{
            printf("%.9f\t", U_T[i]);
        }
    }
    fclose(File);

    File = freopen("Results\\HW3_F_Thomas_cpp.txt", "w", stdout);
    for(int i = 0; i < N; ++i){
        if(i == N-1){
            printf("%.9f", F_T[i]);
        }else{
            printf("%.9f\t", F_T[i]);
        }
    }
    fclose(File);

    return 0;
}
